### PR TITLE
OSSL_PARAM and BIGNUM; ensure at least one byte is allocated

### LINKED ISSUE #20011: OSSL_PARAM_BLD_push_BN of zero secure BIGNUM leads to an OSSL_PARAM with data=NULL and data_size=0
Due to a bug in my own provider, I encountered a strange situation. Not sure if its a bug or works as designed though....

Within my provider export function, I used `OSSL_PARAM_BLD_push_BN()` to push a BIGNUM that was allocated via `BN_secure_new()`, but was never actually set a value (due to a bug in my code), so its value was zero. While additional parameters were pushed to the param-builder, this one was the only with a secure BIGNUM. 

This caused an address sanitizer error when importing these parameters into another provider. The other provider was using `OSSL_PARAM_get_BN` to get the BIGNUM of that parameter.  The error finally occurred in `bin2bn()` (called via `BN_native2bn()` -> `BN_bin2bn()` -> `bin2bn()`) because parameter `s` was NULL and `len` was also zero:

```
    if (endianess == LITTLE) {
        s2 = s + len - 1;
        inc2 = -1;
        inc = 1;
    } else {
        s2 = s;
        inc2 = 1;
        inc = -1;
        s += len - 1;  <----- here we add -1 to NULL which was detected by the address sanitizer
    }
```

It turned out that the OSSL_PARAM created via `OSSL_PARAM_BLD_to_param()` has `data = NULL` and `data_size = 0`. There is no check in `OSSL_PARAM_get_BN()` before passing `data` to `BN_native2bn(`) and also `bin2bn()` does not check if the buffer `s` is NULL or `len` is zero.

However, generating an OSSL_PARAM with a NULL data field by `OSSL_PARAM_BLD_to_param()` is already wrong. 
What happens is the following:

`OSSL_PARAM_BLD_push_BN()` calls `push_BN()` with parameter `sz = NULL`, because `BN_num_bytes(bn)` results in zero bytes for a zero BIGNUM. `push_BN()` then calls `param_push()` again with the zero `sz` and zero `alloc` parameters. Consequently, the `pd->alloc_blocks` is also zero, and it adds zero to `bld->secure_blocks`. If this is the one and only secure parameter, then  `bld->secure_blocks` remains zero. 

`OSSL_PARAM_BLD_to_param()` only allocates a secure buffer for the builder if `bld->secure_blocks` is non-zero. In my case, it is zero, and thus `s` stays NULL.

`param_bld_convert()` then converts the parameters, and uses the passed `secure` pool (which is NULL) to obtain the address of the buffer for the secure parameter:

```
        if (pd->secure) {
            p = secure;
            secure += pd->alloc_blocks;
        } 
```
Since `secure` is NULL, `p` is also NULL. 

The following `BN_bn2nativepad()` would attempt to write to `p`, but since the BIGNUM is zero, it does not output any bytes, and thus ignores that the `to` buffer is NULL (`tolen` is also zero).

As a result we have an OSSL_PARAM with `data = NULL` and `data_size = 0`.

I am aware that it is very unlikely to pass a zero secure BIGNUM via param builder, but maybe its worth to fix this anyway.....
I have no idea how to fix this though, since a zero BIGNUM obviously occupies zero bytes. Maybe we need to special-case this and force it to produce a 1 zero byte OSSL_PARAM for a zero BIGNUM.....

### PR DESCRIPTION
**Pull request by levitte**
A zero BIGNUM contains zero bytes, while OSSL_PARAMs with an INTEGER (or
UNSIGNED INTEGER) data type are expected to have at least one data byte
allocated, containing a zero.  This wasn't handled correctly.

Fixes #20011

### CODE REVIEW

**Comment by slontis** [comment]
@levitte is your understanding here that the BN_new() case is also broken.. i.e. it points to a block for this case, but the next parameter would then reuse this location? The BN_secure case just shows an additional problem of using NULL

**Comment by slontis** [comment]
Would other functions that potentially pass an alloc of zero to param_push() also be a problem
(such as OSSL_PARAM_BLD_push_octet_string() ?)

**Comment by levitte** [comment]
> @levitte is your understanding here that the BN_new() case is also broken.. i.e. it points to a block for this case, but the next parameter would then reuse this location? The BN_secure case just shows an additional problem of using NULL (if there is one param)..

The problem seems to be most visible with `BN_secure`, but that shouldn't matter...  a zero shouldn't become 0 bytes of OSSL_PARAM data, for *any* C number type

**Comment by ifranzki** [comment]
I can confirm that the fix in https://github.com/openssl/openssl/pull/20013/commits/716cb7931bf6ab3f9f984a9187dd7fe74bf9d4ff does fix the problem.

I wonder if the OSSL_PARAM_get_XXX functions should also be made save about invalid OSSL_PARAMs being passed to it. A bogus application could pass invalid OSSL_PARAMs to providers (e.g. via EVP_PKEY_fromdata() ). I guess the OSSL_PARAM_get_XXX functions should not cause any troubles (like address sanitizer findings) if an invalid param is passed to it. I guess passing a OSSL_PARAM to a getter function with data = NULL and/or data_size = 0 does not make any sense no matter what, and should lead to an error return and add an error to the error stack.

**Comment by levitte** [comment]
> I guess passing a OSSL_PARAM to a getter function with data = NULL and/or data_size = 0 does not make any sense no matter what, and should lead to an error return and add an error to the error stack.

Not quite right.  You can do that to query what the minimum data_size should be for a particular OSSL_PARAM key.  This is documented:

https://github.com/openssl/openssl/blob/4b65d79d7132d6e46bfb385a76082f6502ef617b/doc/man3/OSSL_PARAM.pod?plain=1#L113-L116

**Comment by ifranzki** [comment]
> Not quite right. You can do that to query what the minimum data_size should be for a particular OSSL_PARAM key. This is documented:

I know that. I did not say that an OSSL_PARAM can never have data = NULL. What I am saying is that when using with the OSSL_PARAM_get_XXX() functions it can not have data = NULL or data_size = 0. So the OSSL_PARAM_get_XXX() functions should check for that and produce an error return if so.

**Comment by levitte** [comment]
Gah!  For once, I mixed up get and set.  Right you are...

File: test/param_build_test.c [hunk 1]
```diff
@@ -184,6 +260,8 @@ static int template_private_test(int tstid)
         || !TEST_true(OSSL_PARAM_BLD_push_uint32(bld, "i32", 1532))
         || !TEST_true(OSSL_PARAM_BLD_push_uint64(bld, "i64", 9999999))
         || !TEST_true(OSSL_PARAM_BLD_push_size_t(bld, "st", 65537))
+        || !TEST_ptr(zbn = BN_secure_new())
```
Comment (t8m) [inline test/param_build_test.c:263]: You need to free zbn and zbn_res - CI is relevant.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
I added a couple of more commits that have the numeric `OSSL_PARAM_get_XXX()` check that the `data` field is non-NULL and that the `data_size` field is non-zero, and also to have `OSSL_PARAM_set_BN()` ensure that there will be at least one byte of data.

File: test/param_build_test.c (hunk 1 above)
Comment (levitte) [reply test/param_build_test.c:265]: Fixed

File: test/param_build_test.c [hunk 2]
```diff
@@ -38,6 +105,8 @@ static int template_public_test(int tstid)
         || !TEST_true(OSSL_PARAM_BLD_push_int64(bld, "i64", -9999999))
         || !TEST_true(OSSL_PARAM_BLD_push_time_t(bld, "t", 11224))
         || !TEST_true(OSSL_PARAM_BLD_push_double(bld, "d", 1.61803398875))
+        || !TEST_ptr(zbn = BN_new())
```
[the remark this answers is not in the document]
Comment (levitte) [reply test/param_build_test.c:108]: Fixed

**Comment by ifranzki** [comment]
> I added a couple of more commits that have the numeric OSSL_PARAM_get_XXX() check that the data field is non-NULL and that the data_size field is non-zero.

Don't you also need to add the check to `general_get_int()` and `general_get_uint()`? They are called by the `OSSL_PARAM_get_XXX()` functions if the size is not as expected.

Within `general_get_[u]int()` functions `[un]signed_from_[un]signed()` are called and there the `is_negative()` function is called on the `src` which is `p->data` and `p->data_size` which can be NULL. This would produce a segfault as far as I can tell.

**Comment by levitte** [comment]
> > I added a couple of more commits that have the numeric OSSL_PARAM_get_XXX() check that the data field is non-NULL and that the data_size field is non-zero.
> 
> Don't you also need to add the check to `general_get_int()` and `general_get_uint()`? They are called by the `OSSL_PARAM_get_XXX()` functions if the size is not as expected.

I don't understand...  if we have the checks early in `OSSL_PARAM_get_XXX()`, why would I have to reproduce the same checks in `general_get_int()` and `general_get_uint()`?

**Comment by ifranzki** [comment]
> > > I added a couple of more commits that have the numeric OSSL_PARAM_get_XXX() check that the data field is non-NULL and that the data_size field is non-zero.
> > 
> > 
> > Don't you also need to add the check to `general_get_int()` and `general_get_uint()`? They are called by the `OSSL_PARAM_get_XXX()` functions if the size is not as expected.
> 
> I don't understand... if we have the checks early in `OSSL_PARAM_get_XXX()`, why would I have to reproduce the same checks in `general_get_int()` and `general_get_uint()`?

You added the check to `OSSL_PARAM_get_int32()` etc, but not to `OSSL_PARAM_get_int()` etc. Within `OSSL_PARAM_get_int()` it calls `general_get_int()`. So no check before it reaches `general_get_int()`. 

This applies to `OSSL_PARAM_get_int()`, `OSSL_PARAM_get_long()`, `OSSL_PARAM_get_time_t()`, `OSSL_PARAM_get_uint()`, `OSSL_PARAM_get_size_t()`, `OSSL_PARAM_get_ulong()`.

**Comment by levitte** [comment]
The "Provider old versions compat / fips-provider-validation" CI fails...  it appears that the added checks in `OSSL_PARAM_get_XXX()` may break any provider that happens to set `data = NULL` and `data_size = 0` for a zero BN.

That's not promising.  We may have to forego that part of this PR, sadly enough.

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged

master:
b49cf273883c8d3f47542941fe5cc6cb51aec9c9 test/param_build_test.c: test zero BIGNUM
c2ae89148343750e420b72ef1b709ebbc16e47b8 In OSSL_PARAM_set_BN(), make sure that the data_size field is at least 1
c455f87aebf245814ba58d6a398b45ca4e80d1d7 OSSL_PARAM_BLD and BIGNUM; ensure at least one byte is allocated

3.1:
2b7b7eebf9b7d01f8abab88b9ae0685f33c74b54 test/param_build_test.c: test zero BIGNUM
fcc224a3c5c868ee4c0f5ab71d2f47e7f620bc98 In OSSL_PARAM_set_BN(), make sure that the data_size field is at least 1
e33c37aead7e7dd5647d949db7be875c8fa8c137 OSSL_PARAM_BLD and BIGNUM; ensure at least one byte is allocated

3.0:
5601648e91d99696bc7ae37e8972c5b4b15cf832 test/param_build_test.c: test zero BIGNUM
f51b4ebb07948d942cd6536fe48e325655349d97 In OSSL_PARAM_set_BN(), make sure that the data_size field is at least 1
1932e595c8061bcb8d2718eaf5fdb6038bdec631 OSSL_PARAM_BLD and BIGNUM; ensure at least one byte is allocated

**Pull request closed without merging**